patches/server/0969-New-specific-event-watching-command.patch

```diff
@@ -0,0 +1,199 @@
++import static net.kyori.adventure.text.format.NamedTextColor.GREEN;
++import static net.kyori.adventure.text.format.NamedTextColor.RED;
++
++@DefaultQualifier(NonNull.class)
++public class EventWatcherCommand implements PaperSubcommand {
++
++    private static EventWatcherCommand instance; //will be null if called before command reg
++    /**
++     * The key is the equivalent of {@link Class#getName()} of your event class, and the value is our list of {@link Wrapper}'s
++     */
++    private final Map<String, List<Wrapper>> watching = new ConcurrentHashMap<>(); //only concurrent in the rare case that multiple people are debugging, should be unnecessary
++    private final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd_HH.mm.ss");
++    private final Path PARENT = Path.of("event-debug");
++    private EventWatcherCommand() {
++    }
++
++    public static synchronized EventWatcherCommand get() {
++        if (instance == null)
++            instance = new EventWatcherCommand();
++        return instance;
++    }
++
++    public void tryLog(Event event, @Nullable Listener listener, @Nullable EventExecutor executor) { //listener should only be null on the initial event call
++        List<Wrapper> wrappers = watching.get(event.getClass().getName());
++        if (wrappers != null)
++            wrappers.add(new Wrapper(event, listener, executor));
++    }
++
++    @Override
++    public boolean execute(CommandSender sender, String subCommand, String[] args) {
++        if (args.length < 1) {
++            sender.sendMessage(text("Invalid Usage! /paper watchevent <event name>", RED));
++            return true;
++        }
++        List<Wrapper> wrappers = watching.get(args[0]);
++        if (wrappers != null) {
++            try {
++                dumpToFile(sender, args[0], wrappers);
++            } catch (IOException ex) {
++                sender.sendMessage(text("Failed to log event information! See the console for more info.", RED));
++                MinecraftServer.LOGGER.warn("Error occurred while logging event information", ex);
++            }
++        } else {
++            watching.put(args[0], new ArrayList<>());
++            sender.sendMessage(text("Now watching event " + args[0] + "!", GREEN));
++        }
++        return true;
++    }
++
++    @Override
++    public List<String> tabComplete(CommandSender sender, String subCommand, String[] args) {
++        return switch (args.length) {
++            case 0 -> new ArrayList<>(DumpListenersCommand.eventClassNames());
++            case 1 -> DumpListenersCommand.eventClassNames().stream()
++                .filter(clazz -> clazz.toLowerCase(Locale.ROOT).contains(args[0].toLowerCase(Locale.ROOT)))
++                .toList();
++            default -> Collections.emptyList();
++        };
++    }
++
++    private void dumpToFile(CommandSender sender, String event, List<Wrapper> wrappers) throws IOException {
```

Comment:
In general, events can quickly accumulate. I think it may be better to instead actively log to a file async, similar to the latest.log that is auto generated. Have it log an event instead of having to store all of these event calls in memory.

Reply:
Is latest.log not just sl4j? Would appreciate a point towards an async internal task method